### DISCUSSION TITLE
Trimming android application

### DISCUSSION THREAD

**Discussion by JenyaRostov** [AvaloniaUI/Avalonia#13119] [Q&A] [answered]
I have created an Avalonia.Xplat application and i'm trying to publish it trimmed
I'm using these .csproj properties on my  Android csproject

PublishTrimmed - true
 TrimMode - full
    OptimizationPreference - Size
RuntimeIdentifier - android-arm64

I have tried TrimMode - copyused that i found in some github issue but all i got is a lot of weird compilation errors
My ultimate goal is to reduce final application size, speed doesn't matter
Any tips?

**Comment by maxkatz6** [comment]
> TrimMode - copyused

It won't get you the best trimming results. "full" is what you need.

Read MS documentation on this topic: https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0

It's mostly standardized for any .NET 7 app.

> all i got is a lot of weird compilation errors

We can't help you without any information provided.

**Comment by timunie** [comment]
Compiled Bindings should be enabled. Make sure to use no reflection if possible.

**Comment by JenyaRostov** [comment]
![IMG_20231004_111750_734](https://github.com/AvaloniaUI/Avalonia/assets/52300935/8dea0db0-13e3-44b8-ac8d-a1113376b40b)
Those are the errors I've got from logcat after trying to run the app wil TrimMode full
No bindings and reflection are used
Do i still have to enable compiled bindings?

**Reply by maxkatz6** [reply]
Do you use AccentColors.xaml anywhere? 
You need to include FluentTheme instead.

Also, what Avalonia version are you using?

**Reply by JenyaRostov** [reply]
No, i don't use it
I'm now using FluentTheme
Version is 11.0.4
I was using preview before and updating to 11.0.4 fixed the problem
But my application still has a size of 166mb
113 mb of which is "User data"

**Reply by JenyaRostov** [reply]
Separate issue but after updating to 11.0.4 i started getting this weird error trying to run my application on either emulator or my phone
<img width="826" alt="rider64_sBfubCEv8n" src="https://github.com/AvaloniaUI/Avalonia/assets/52300935/3822d13a-3e46-470e-b01d-6d63dedfdefe">(because it is being used by another process.)
I'm using rider

**Reply by maxkatz6** [reply]
> But my application still has a size of 166mb

It is way higher than minimal Avalonia application, so you would need to investigate what heavy dependencies are you using.

You also can build apk files only for specific architectures, like specifying `<RuntimeIdentifiers>android-arm64;android-arm</RuntimeIdentifiers>` in csproj (disabling x86/x64 support, which isn't common on mobile).

**Reply by maxkatz6** [reply]
> I was using preview 

Actually, you might need to recreate base projects from the "avalonia.xplat" template, to keep it closer to defaults. It won't affect much though.

**Reply by JenyaRostov** [reply]
Okay, will try with minimal app 
but shouldn't these dependencies contribute to "App size" and not "User data" which appears only after the very first launch of the app?

**Reply by timunie** [reply, marked as answer]
I don't think Avalonia adds much user data, if even any? Check with a minimal sample and also double-check your other dependencies. Trimming indeed control App size, not user data size.

**Reply by JenyaRostov** [reply]
Minimal sample looks fine in terms of storage, 44mb
looks like it's a skill issue on my part, will try to figure it out
thanks!
